File: 配套资源/配套代码/程序设计竞赛专题挑战教程-代码/第6章合根植物.py
#合根植物  2017年第八届决赛  lanqiaoOJ 110
def find_set(x):                #有路径压缩优化的查询
    if(x != s[x]):  s[x] = find_set(s[x])   
    return s[x]
def merge_set(x, y):
    x = find_set(x); y = find_set(y)
    if x == y:   return False
    s[y] = x
    return True
m, n = map(int, input().split())
k = int(input())
s = list(range(m*n+1))       #并查集，初始化
ans = m * n
for i in range(k):
    x, y = map(int, input().split())
    if merge_set(x, y):   ans -= 1
print(ans)

File: 配套资源/配套代码/程序设计竞赛专题挑战教程-代码/第6章合根植物.cpp
//合根植物  2017年第八届决赛  lanqiaoOJ 110
#include<bits/stdc++.h>
using namespace std;
const int N = 1e6+10;
int s[N];
int find_set(int x){
    if(x!=s[x])  s[x]=find_set(s[x]);
    return s[x];
}
int main(){
    int n,m,k;   scanf("%d%d%d",&n,&m,&k);
    for(int i=1;i<=N;i++)  s[i]=i;   //初始化并查集
    while(k--){
        int a,b;   scanf("%d%d",&a,&b);
        int pa = find_set(a),pb=find_set(b);
        if(pa!=pb)  s[pa] = pb;    //合并并查集
    }
    int ans=0;
    for(int i=1;i<=n*m;i++)
        if(s[i]==i)  ans++;
    printf("%d",ans);
    return 0;
}

